core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -744,7 +744,9 @@ object SparkSubmit extends CommandLineUtils with Logging {
     }
 
     // Ignore invalid spark.driver.host in cluster modes.
-    if (deployMode == CLUSTER) {
+    if (isYarnCluster) {
+      sparkConf.set("spark.driver.host", "${env:NM_HOST}")
```

Comment:
This actually requires `spark.sql.variable.substitute` set to `true`, though by default it is, if some user accidentally changes it, the code won't work.

Reply:
Good to know. It seems that substitute is unconditional in core for this `ConfigEntryWithDefaultString`. I liked the brevity of this approach. Pending @vanzin's feedback I can redo it in explicit code again.

Reply:
re: sql, this is not a sql conf object, so sql options have no effect.

I think it would be better to have this in YARN's `Client.scala` instead, in the spirit of not adding more backend-specific logic to this class.